### PR TITLE
Issue_5426: Add default antiAffinity example to Kafka

### LINKED ISSUE #5426: Kafka & Zookeeper Stateful Sets don't use anti-affinities by default (resulting in pod-node overlap)
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG REPORT

**Version of Helm and Kubernetes**:
Helm v2.9.0
Kubernetes 1.9.6

**Which chart**:
Kafka & ZooKeeper

**What happened**:
There are no anti-affinities defined on either Kafka or Zookeeper, which is problematic, since I end up having multiple Zookeeper and/or Kafka instances on the same node.

**What you expected to happen**:
I'd expect that by default all Kafka and Zookeeper instances would end up on different nodes (by making use of anti-affinities).

**How to reproduce it** (as minimally and precisely as possible):
Just install the default Kafka Helm Chart.

**Anything else we need to know**:
No

### PR DESCRIPTION
**Pull request by alexvicegrab**
**What this PR does / why we need it**:

Added a default antiAffinity to Kafka and Zookeeper, since it's unlikely we will want to have them run on the same node, and these can be expanded with more specific affinities & antiAffinities.

Also fix an error in the Zookeeper chart that suggests there exists an antiAffinity (soft/hard) flag, despite this not being used anywhere in the templates.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes #5426 

**Special notes for your reviewer**:

All charts lint correctly and launch to Running on my AKS 3 node cluster with the (new) default values). Also, running:

```
kubectl get pods -n ${K8S_NAMESPACE} -o wide | grep ${RELEASE_NAME}
```

shows all pods running on different nodes.

@benjigoldberg
@lachie83
@kow3ns

### CODE REVIEW

**Comment by alexvicegrab** [comment]
Finished the CLA

**Comment by benjigoldberg** [comment]
@alexvicegrab I'm not really crazy about the idea of adding a default anti-affinity policy. You will actually find that there is more controversy on this topic than you might think. In our cluster we run our anti-affinity rules very much as you've defaulted to here. That being said, if you take a look at this [Kubecon presentation from 2017 in Austin, TX](http://events17.linuxfoundation.org/events/kubecon-and-cloudnativecon-north-america/program/schedule) the presenter makes a good argument that it actually makes sense to colocate Zookeeper and Kafka to reduce network overhead.

My preference would be to define something in commented format, much like the Zookeeper example. We can then of course link to other references and documentation that encourage the reader to come to a conclusion that best fits their use case.

What do you think @alexvicegrab?

**Comment by alexvicegrab** [comment]
Hi @benjigoldberg, your comments are fair.

I did notice that the standard chart did not set up any affinity rules, but I think in very many cases people will want to implement some anti-affinity to keep the nodes separate and crash tolerant. Hence I can add a commit to comment the additions out and restore the default `{}`, but make it easy for individuals to add this feature.

To be fair, my intention was to collocate the zookeeper nodes and the Kafka nodes (i.e. having each node have one Kafka node and one zookeeper node), this would probably need an added affinity rule for the Kafka nodes to attach themselves to nodes that run Zookeeper. I can create an extra (commented out) rule like this and add it also as an example.

So in short:
1) comment out the new affinity rules and keep them as examples.
2) add an extra affinity rule example that uses positive affinity to get Kafka nodes to collocate with nodes running Zookeeper.

Would this be agreeable?

File: incubator/kafka/values.yaml [hunk 1]
```diff
@@ -272,7 +296,7 @@ zookeeper:
   persistence:
     enabled: false
     ## The amount of PV storage allocated to each Zookeeper pod in the statefulset
-    # storage: "2Gi"
+    # size: "2Gi"
```
Comment (alexvicegrab) [inline incubator/kafka/values.yaml:299]: This key was incorrectly specified in the original values.yaml, should be "size", not "storage"

File: incubator/zookeeper/README.md [hunk 2]
```diff
@@ -112,7 +112,7 @@ Spreading allows you specify an anti-affinity between ZooKeeper servers in the e
 | Parameter | Description | Default |
 | --------- | ----------- | ------- |
 | `schedulerName` | Name of scheduler to use (other than the default). | `nil` |
-| `antiAffinity` | If present it must take the values 'hard' or 'soft'. 'hard' will cause the Kubernetes scheduler to not schedule the Pods on the same physical node under any circumstances 'soft' will cause the Kubernetes scheduler to make a best effort to not co-locate the Pods, but, if the only available resources are on the same node, the scheduler will co-locate them. | `hard` |
```
Comment (alexvicegrab) [inline incubator/zookeeper/README.md]: "antiAffinity" does not appear to get used anywhere in the templates, instead we define "affinity"

File: incubator/kafka/templates/service-brokers.yaml [hunk 3]
```diff
@@ -18,19 +18,19 @@ spec:
...
-    release: {{ .Release.Name | quote }}
+    app: {{ include "kafka.name" . }}
+    release: {{ .Release.Name }}
 ---
 {{- if and .Values.prometheus.kafka.enabled .Values.prometheus.operator.enabled }}
 apiVersion: v1
 kind: Service
 metadata:
-  name: {{ template "kafka.fullname" . }}-exporter
+  name: {{ include "kafka.fullname" . }}-exporter
```
Comment (alexvicegrab) [inline incubator/kafka/templates/service-brokers.yaml:28]: e.g. updating chart to consistently use include over template (rather than intermix the two), as described here:

https://github.com/kubernetes/helm/blob/master/docs/chart_template_guide/named_templates.md

File: incubator/kafka/values.yaml (hunk 1 above)
Comment (benjigoldberg) [reply incubator/kafka/values.yaml:299]: 👍

File: incubator/kafka/templates/service-brokers.yaml (hunk 3 above)
Comment (benjigoldberg) [reply incubator/kafka/templates/service-brokers.yaml:28]: 👍

**Comment by benjigoldberg** [comment]
@alexvicegrab That plan sounds good to me. I think adding examples could be a useful way to give people advice on how to structure their affinity rules.

Also, I didnt notice this on my first pass -- can you separate the zookeeper changes into a separate PR? We tend to try and keep PRs contained to one chart at a time. I also don't have merge permissions on the Zookeeper chart, so we'll need to pull somebody from the `OWNERS.MD` in that chart to sign off on this as well if we dont split them out.

**Comment by alexvicegrab** [comment]
Hi Benji,

I will create a separate branch where I keep all the zookeeper changes and remove them from this branch, so you can merge, please bear with me.

Thanks!
S.

**Comment by alexvicegrab** [comment]
@benjigoldberg, I have removed changes to Zookeeper, please review.

**Comment by benjigoldberg** [comment]
@alexvicegrab thanks for the contribution!

**Comment by alexvicegrab** [comment]
@benjigoldberg, Thank you, looks like the pull-charts-e2e works, but the tide test can't proceed until the "needs-ok-to-test" label, could you add an /ok-to-test?

**Comment by benjigoldberg** [comment]
hah! Sorry @alexvicegrab my bad.

/ok-to-test
